torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -547,6 +560,246 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
+  //   -   [8]: nnz
+  //
+  class SparseTensorMetadata {
+   public:
+    static constexpr auto dim = 9;
+
+    // Construct from an existing metadata tensor to facilitate structured
+    // access to metadata from peers, after gathering it.
+    explicit SparseTensorMetadata(at::Tensor metadata)
+        : metadata_(metadata), data_(metadata_.data<long>()) {
+      AT_ASSERT(metadata.scalar_type() == at::kLong);
+      AT_ASSERT(metadata.dim() == 1);
+      AT_ASSERT(metadata.size(0) == dim);
+    }
+
+    // Populate the metadata.
+    void populate_from_sparse_tensor(const at::Tensor& tensor) {
+      const auto sparse_dim = tensor.sparse_dim();
+      AT_ASSERT(sparse_dim <= 4);
+      for (auto i = 0; i < 4; i++) {
+        if (i < sparse_dim) {
+          data_[i] = tensor.size(i);
+        }
+      }
+      const auto dense_dim = tensor.dense_dim();
+      AT_ASSERT(dense_dim <= 4);
+      for (auto i = 0; i < 4; i++) {
+        if (i < dense_dim) {
+          data_[i + 4] = tensor.size(sparse_dim + i);
+        }
+      }
+      data_[8] = tensor._nnz();
+    }
+
+    std::vector<int64_t> sizes() const {
+      std::vector<int64_t> sizes;
+      // Sparse sizes
+      for (auto i = 0; i < 4; i++) {
+        if (data_[i] <= 0) {
+          break;
+        }
+        sizes.push_back(data_[i]);
+      }
+      // Dense sizes
+      for (auto i = 4; i < 8; i++) {
+        if (data_[i] <= 0) {
+          break;
+        }
+        sizes.push_back(data_[i]);
+      }
+      return sizes;
+    }
+
+    int64_t nnz() const {
+      return data_[8];
+    }
+
+   protected:
+    at::Tensor metadata_;
+    long* data_;
+  };
+
+  // Sparse allreduce is implemented with allgather on indices and values.
+  // Every process then sums the resulting sparse tensors locally.
+  // The nnz for sparse tensors may be different across processes, so first
+  // we run allgather on the nnz, and then allgather with max(nnz).
+  // We could use an allgatherv for this, if it were available.
+  at::Tensor allreduce(std::vector<at::Tensor>& tensors) {
+    auto input = tensors[0];
+
+    // Perform local reduction if we have multiple inputs.
+    for (size_t i = 1; i < tensors.size(); i++) {
+      input += tensors[i];
+    }
+
+    // Need to coalesce before we can access indices and values.
+    input = input.coalesce();
+
+    // Gather metadata information from all ranks.
+    auto metadata = allgather_metadata(input);
+
+    // Sanity check dimensionality across ranks.
+    {
+      const auto expected = metadata[context->rank].sizes();
+      for (size_t i = 0; i < context->size; i++) {
+        if (i == context->rank) {
+          continue;
+        }
+        const auto actual = metadata[i].sizes();
+        AT_CHECK(actual == expected, "Sparse dimensions do not match");
+      }
+    }
+
+    // Gather all indices and all values.
+    auto indices = allgather_indices(input, metadata);
+    auto values = allgather_values(input, metadata);
```

Comment:
should these two be done in one allgather?

Reply:
This should be possible in theory, since we have all data. But the data types may be different (the indices are always `long` and the values may be anything). To combine them in a single call we'd need to cast both of them to `char` buffers since it would run on the byte level.